### PR TITLE
ADD fips assessment test for openshift-serverless

### LINKED ISSUE #11: Implement openshift-serverless testing with opcap-ansible
The steps to implement the testing are:
- Create the tasks according to the check.txt
- Handle any necessary yaml
- Generate results to be reported back to opcap-ansible

opcap-ansible integration will be treated on its own issue and repo.

### PR DESCRIPTION
**Pull request by acmenezes**
Added tasks/main.yaml to be used by opcap-ansible and run the tests.

fixes #11

### CODE REVIEW

**Comment by madorn** [comment]
@acmenezes thanks for the PR.

I understand that cleanup of the cloned files and deletion of the namespace exists outside this file.

Should we add additional tasks to this particular file to remove its created CRs/CRDs?

**Comment by acmenezes** [comment]
> @acmenezes thanks for the PR.
> 
> I understand that cleanup of the cloned files and deletion of the namespace exists outside this file.
> 
> Should we add additional tasks to this particular file to remove its created CRs/CRDs?

@madorn , I've included the tasks for deleting the CRs. The CRDs though are created and managed by the operator. I believe the cleanup task on the opcap-ansible side will delete the operator subscription. With that anything, including CRDs that are in the bundle, would be deleted.

**Comment by madorn** [comment]
Thanks @acmenezes 

Runs fine now. As discussed, I am following up on why `absent` is only confirming `metadata.DeletionTimestamp` so objects may still stick around if they have a finalizer that has not been removed.

Researching easiest way to confirm a `GET` on that object returns nil.

**Comment by madorn** [comment]
@acmenezes we can add `wait: true` to addition to `state: absent` to ensure object is gone before proceeding.

**Comment by madorn** [comment]
I was able to confirm successful run with `cleanup: true` set in vars

File: openshift-serverless/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    state: absent
+    wait: true
+- name: Delete knative eventing
+  kubernetes.core.k8s:
+    name: knative-eventing
+    namespace: knative-eventing
+    api_version: operator.knative.dev/v1beta1
+    kind: KnativeEventing
+    state: absent
+    wait: true
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:204]: Need LF

File: openshift-serverless/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        labels:
+          opdev: opcap
+    state: present
+- name: Test knative serving conditions type Ready
+  kubernetes.core.k8s_info:
+    namespace: knative-serving
+    name: knative-serving
+    api_version: operator.knative.dev/v1beta1
+    kind: knativeserving
+    wait: yes
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:45]: Should be 'true'.

File: openshift-serverless/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                  - name: RESPONSE
+                    value: "Hello Serverless!"            
+    state: present
+- name: Get hello service Conditions "Ready"
+  kubernetes.core.k8s_info:
+    namespace: knative-serving
+    name: hello
+    api_version: serving.knative.dev/v1
+    kind: Service
+    wait: yes
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:78]: Should be 'true'.

File: openshift-serverless/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      status: "True"
+      type: "Ready"
+    wait_timeout: 90
+- name: Get hello service Conditions "RoutesReady"
+  kubernetes.core.k8s_info:
+    namespace: knative-serving
+    name: hello
+    api_version: serving.knative.dev/v1
+    kind: Service
+    wait: yes
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:89]: Should be 'true'.

File: openshift-serverless/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+      status: "True"
+      type: "RoutesReady"
+    wait_timeout: 90
+  register: service
+  failed_when: 
+    - (service.resources[0].status.conditions | selectattr('type', 'search', 'Ready') | list | last).status != "True"
+    - (service.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True"
+- name: Check hello service - must return hello serverless
+  ansible.builtin.uri:
+    url: "{{service.resources[0].status.url}}"
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:100]: Jinja spacing should be more like this for readability:

```suggestion
    url: "{{ service.resources[0].status.url }}"
```

File: openshift-serverless/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    name: hello
+    api_version: serving.knative.dev/v1
+    kind: Service
+    wait: yes
+    wait_condition:
+      status: "True"
+      type: "RoutesReady"
+    wait_timeout: 90
+  register: service
+  failed_when:
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:95]: Trailing spaces.

File: openshift-serverless/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                env:
+                  - name: RESPONSE
+                    value: "Hello Serverless!"
+- name: Get hello-autoscale service Conditions "Ready"
+  kubernetes.core.k8s_info:
+    namespace: knative-serving
+    name: hello-autoscale
+    api_version: serving.knative.dev/v1
+    kind: Service
+    wait: yes
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:131]: Should be 'true'.

File: openshift-serverless/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      status: "True"
+      type: "Ready"
+    wait_timeout: 90
+- name: Get hello-autoscale service Conditions "RoutesReady"
+  kubernetes.core.k8s_info:
+    namespace: knative-serving
+    name: hello-autoscale
+    api_version: serving.knative.dev/v1
+    kind: Service
+    wait: yes
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:142]: Should be 'true'.

File: openshift-serverless/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    wait_condition:
+      status: "True"
+      type: "RoutesReady"
+    wait_timeout: 90
+  register: service_autoscale
+  failed_when:
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:148]: Trailing spaces.

File: openshift-serverless/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+      status: "True"
+      type: "RoutesReady"
+    wait_timeout: 90
+  register: service_autoscale
+  failed_when: 
+    - (service_autoscale.resources[0].status.conditions | selectattr('type', 'search', 'Ready') | list | last).status != "True"
+    - (service_autoscale.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True"
+- name: Check hello service autoscale - must return Hello world from sample HTTP server
+  ansible.builtin.uri:
+    url: "{{service_autoscale.resources[0].status.url}}"
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:153]:
```suggestion
    url: "{{ service_autoscale.resources[0].status.url }}"
```

File: openshift-serverless/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    return_content: true
+    validate_certs: false
+  register: autoscale_hello_msg
+  failed_when: "'Hello world' not in autoscale_hello_msg.content"
+# If it didn't failed till here all tests must have succeeded
+- name: Openshift-Serverless results
+  debug:
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:160]: Always use the FQCN

`ansible.builtin.debug`

File: openshift-serverless/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    name: hello-autoscale
+    namespace: knative-serving
+    state: absent
+    wait: true
+- name: Delete knative hello service
+  kubernetes.core.k8s:
+    name: hello
+    namespace: knative-serving
+    api_version: serving.knative.dev/v1
+    kind: Service
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:186]: Trailing spaces.

File: openshift-serverless/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    - (service_autoscale.resources[0].status.conditions | selectattr('type', 'search', 'Ready') | list | last).status != "True"
+    - (service_autoscale.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True"
+- name: Check hello service autoscale - must return Hello world from sample HTTP server
+  ansible.builtin.uri:
+    url: "{{service_autoscale.resources[0].status.url}}"
+    return_content: true
+    validate_certs: false
+  register: autoscale_hello_msg
+  failed_when: "'Hello world' not in autoscale_hello_msg.content"
+# If it didn't failed till here all tests must have succeeded
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:158]: Good first cut. We will want to add some error handling via blocks to make sure we get to the cleanup phase.

File: openshift-serverless/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    kind: Service
+    wait: yes
+    wait_condition:
+      status: "True"
+      type: "RoutesReady"
+    wait_timeout: 90
+  register: service
+  failed_when: 
+    - (service.resources[0].status.conditions | selectattr('type', 'search', 'Ready') | list | last).status != "True"
+    - (service.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True"
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:97]: I think you want this to be failed if either of these occur. failed_when: with a list has an implicit `and`. 

If you want `or`, you will need to use it explicitly.

```suggestion
    failed_when: >
        ((service.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True") or
        ((service.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True")
```

File: openshift-serverless/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    kind: Service
+    wait: yes
+    wait_condition:
+      status: "True"
+      type: "RoutesReady"
+    wait_timeout: 90
+  register: service_autoscale
+  failed_when: 
+    - (service_autoscale.resources[0].status.conditions | selectattr('type', 'search', 'Ready') | list | last).status != "True"
+    - (service_autoscale.resources[0].status.conditions | selectattr('type', 'search', 'RoutesReady') | list | last).status != "True"
```
Comment (bcrochet) [inline openshift-serverless/tasks/main.yml:150]: Same comment as above.

**Review by bcrochet — CHANGES_REQUESTED** [review]
A couple more observations.

File: openshift-serverless/tasks/main.yml (hunk 1 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+        labels:
+          opdev: opcap
+    state: present
+- name: Test knative eventing conditions type Ready
+  kubernetes.core.k8s_info:
+    namespace: knative-eventing
+    name: knative-eventing
+    api_version: operator.knative.dev/v1beta1
+    kind: knativeeventing
+    wait: yes
```
[the remark this answers is not in the document]
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml (hunk 2 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+    name: knative-serving
+    api_version: operator.knative.dev/v1beta1
+    kind: knativeserving
+    wait: yes
+    wait_condition:
+      status: "True"
+      type: "Ready"
+    wait_timeout: 90
+  register: knative_serving
+  failed_when: (knative_serving.resources[0].status.conditions | selectattr('type', 'search', 'Ready') | list | last).status != "True"
```
[the remark this answers is not in the document]
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+        labels:
+          opdev: opcap
+      spec:
+        template:
+          spec:
+            containers:
+              - image: docker.io/openshift/hello-openshift
+                env:
+                  - name: RESPONSE
+                    value: "Hello Serverless!"
```
[the remark this answers is not in the document]
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml (hunk 9 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml (hunk 10 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml (hunk 11 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml (hunk 12 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: done

File: openshift-serverless/tasks/main.yml (hunk 13 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml:158]: @bcrochet ,  what kind of error handling are you suggesting here? Do you have any examples? Ansible is supposed to be idempotent which means if I have an error I can fix it and run it again until I get to the clean up phase. I would disagree a bit here and maybe would suggest to put the cleanup tasks in a separate file for people to run it separately if that's needed after a failed attempt. I don't think we should attempt to create too much logic here. It's supposed to be simple and straightforward. I'm afraid we'll fallback to Go if we need to really get some good logic in. But let me know if you have any working example on this and what you think.

Comment (bcrochet) [reply openshift-serverless/tasks/main.yml:158]: This is what I'm speaking of: 
https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_blocks.html#handling-errors-with-blocks

Of course, there are other ways of making sure that even if something fails, it doesn't stop the processing of other operators for instance. But again, we can visit it later.

File: openshift-serverless/tasks/main.yml (hunk 14 above)
Comment (bcrochet) [reply openshift-serverless/tasks/main.yml:97]: Or they should be on the corresponding tasks.

File: openshift-serverless/tasks/main.yml (hunk 13 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: Hi @bcrochet , thanks for the suggestion. I've implemented it and also changed the name of the directory to match the package name. Just pulled your PR on opcap-ansible dynamic and it works with this one.

File: openshift-serverless/tasks/main.yml (hunk 15 above)
Comment (acmenezes) [reply openshift-serverless/tasks/main.yml]: Done!

**Pull request merged**